src/main/java/org/eclipse/uprotocol/streamer/UStreamer.java

```diff
@@ -0,0 +1,183 @@
+        TransportListener listener = new TransportListener(in, out);
+        UUri out_uri = UUri.newBuilder().setAuthority(out.getAuthority()).build();
+        UStatus result = in.getTransport().registerListener(out_uri, listener);
+        if (result.getCode() != UCode.OK) {
+            return result;
+        }
+
+        listeners.add(listener);
+
+
+        // Published messages do not contain a sink so we cannot setup simple static forwarding rules based
+        // on the sink address like notifications, requests and responses. Registering a listener for all
+        // published events would mean that the streamer gets each and ever single published topic on the IN
+        // Route (not what we want). What we need to know is what published topics does the streamer really 
+        // need to care about (i.e register a listener for) so we don't get everything, this is only published
+        // topics that there is a subscriber in OUT UAuthority that has subscribed to said topic.
+        //
+        // The good thing is that the subscribers information (URI) is recorded in usubscription service when 
+        // the subscriber calls subscribe(), all we need to do is fetch ths list of subscriptions that match the 
+        // IN UAuthority when the subscriber is the OUT UAuthority and we have the magic list we need!
+        FetchSubscriptionsRequest request = FetchSubscriptionsRequest.newBuilder()
+            .setTopic(UUri.newBuilder().setAuthority(in.getAuthority()))
+            .setSubscriber(SubscriberInfo.newBuilder().setUri(UUri.newBuilder().setAuthority(out.getAuthority())))
+            .build();
+
+        // Get the list of subscriptions
+        FetchSubscriptionsResponse response = submgr.fetchSubscriptions(request);
+        
+        // Make the streamer register for the published events that the 
+        // OUT UAuthority wanted 
+        response.getSubscriptionsList().forEach(p -> {
+            in.getTransport().registerListener(p.getTopic(), listener);
+        });
+
+        // Any changes to subscription will require dynamic add/removal of listeners
+        return result;
+    }
+
+    
+    /**
+    * Remove a forwarding rule from the streamer. 
+    * 
+    * @param in input {@code Route} that the streamer listens to
+    * @param out output {@code Route} that the streamer forwards the message to
+    * @return {@code UStatus} with UCode.OK if the forwarding rule was removed successfully,
+    *         {@code UStatus} with UCode.NOT_FOUND if the forwarding rule was not found,
+    *         {@code UStatus} with UCode.INVALID_ARGUMENT if the input and output routes are the same
+    */
+    public UStatus deleteForwardingRule(Route in, Route out) {
+        Objects.requireNonNull(in, "input cannot be null.");
+        Objects.requireNonNull(out, "output cannot be null.");
+        
+        if (in.equals(out)) {
+            return UStatus.newBuilder().setCode(UCode.INVALID_ARGUMENT).build();
```

Comment:
who cares? if you can't delete, if there is nothing to delete, it is fine.
delete means I dont want something to be there anymore. If it is not possible then just say ok, deleted.
IMHO, you only give an error when something is there and you could not delete it, meaning I wanted it gone and it is still there. All other cases are, ok - it is not there anymore.
This is from a developer perspective.